### PR TITLE
Cluster autoscaler doc - part 1

### PR DESCRIPTION
**Pull request by mwielgus**
cc: @piosz @jszczepkowski @fgrzadkowski

### CODE REVIEW

File: docs/admin/cluster-management.md [hunk 1]
```diff
@@ -66,37 +66,43 @@ Instance Group will take care of putting appropriate image on new machines and s
...
-the autoscaler will choose the bigger number. The number of nodes in the cluster set by the autoscaler will be limited from `KUBE_AUTOSCALER_MIN_NODES` (default: 1)
-to `KUBE_AUTOSCALER_MAX_NODES` (default: the initial number of nodes in the cluster).
+Cluster autoscaler also scales down the cluster if it notices that some node is not needed anymore in a sence
+that:
 
-The autoscaler is implemented as a Compute Engine Autoscaler.
-The initial values of the autoscaler parameters set by `kube-up.sh` and some more advanced options can be tweaked on
-`Compute > Compute Engine > Instance groups > your group > Edit group`[Google Cloud Console page](https://console.developers.google.com)
-or using gcloud CLI:
+* It has more that 50% of cpu and memory not requested by any pod.
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: I'd remove this part. It's an implenentation details, which is a pure heuristic to find good candidates for removal.

File: docs/admin/cluster-management.md [hunk 2]
```diff
@@ -66,37 +66,43 @@ Instance Group will take care of putting appropriate image on new machines and s
...
+* All of its pods can be easily restarted on the other nodes (they are replciated, don't have local storage) and there
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: I think you need to explicitly say that this node is not needed for an extended period of time (currently 10 min, but this may change).

File: docs/admin/cluster-management.md [hunk 3]
```diff
@@ -66,37 +66,43 @@ Instance Group will take care of putting appropriate image on new machines and s
...
-`Compute > Compute Engine > Instance groups > your group > Edit group`[Google Cloud Console page](https://console.developers.google.com)
-or using gcloud CLI:
+* It has more that 50% of cpu and memory not requested by any pod.
+* All of its pods can be easily restarted on the other nodes (they are replciated, don't have local storage) and there 
+  is no important system pods among them.
 
-```shell
-gcloud alpha compute autoscaler --zone $ZONE <command>
-```
+Cluster autoscaler is configured per instance group(GCE) or node pool(GKE).
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: Spaces before `(`

File: docs/admin/cluster-management.md [hunk 4]
```diff
@@ -66,37 +66,43 @@ Instance Group will take care of putting appropriate image on new machines and s
...
+* All of its pods can be easily restarted on the other nodes (they are replciated, don't have local storage) and there 
+  is no important system pods among them.
 
-```shell
-gcloud alpha compute autoscaler --zone $ZONE <command>
-```
+Cluster autoscaler is configured per instance group(GCE) or node pool(GKE). 
+
+If you are using GCE then you can either enable it while creating a cluster with kube-up.sh script or later with
+kube-push.sh. To configure cluser autoscaler you have to set 3 environment variables:
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: Does `kube-push.sh` actually work? I don't think we've ever tested it.

File: docs/admin/cluster-management.md [hunk 5]
```diff
@@ -66,37 +66,43 @@ Instance Group will take care of putting appropriate image on new machines and s
...
+
+* `KUBE_ENABLE_CLUSTER_AUTOSCALER` - it enables cluster autoscaler if set to true.
+* `KUBE_AUTOSCALING_MIN_NODES` - minimum number of nodes in the cluster.
+* `KUBE_AUTOSCALING_MAX_NODES` - maximum number of nodes in the cluster.
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: Can you give an example of kube-up command with those 3 flags?

File: docs/admin/cluster-management.md [hunk 6]
```diff
@@ -66,37 +66,43 @@ Instance Group will take care of putting appropriate image on new machines and s
...
+
+On GKE you configure cluster autoscaler either on cluster creation or update or when creating a particular node pool
+(which you want to be autoscaled) by passing flags `--enable-autoscaling` `--min-nodes` and `--max-nodes`
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: Can you give example of gcloud commands? Or a link to gcloud documentation?

File: docs/admin/cluster-management.md [hunk 7]
```diff
@@ -66,37 +66,43 @@ Instance Group will take care of putting appropriate image on new machines and s
...
+to the corresponding `gcloud` commands.
 
-Note that autoscaling will work properly only if node metrics are accessible in Google Cloud Monitoring.
-To make the metrics accessible, you need to create your cluster with `KUBE_ENABLE_CLUSTER_MONITORING`
-equal to `google` or `googleinfluxdb` (`googleinfluxdb` is the default value). Please also make sure
-that you have Google Cloud Monitoring API enabled in Google Developer Console.
+** Cluster autoscaler assumes that all nodes inside a signle instance group or node pool are the roughly the same,
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: What do you mean by "roughly the same"?

Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: Remove "the" before "roughly"

Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: When viewing md doc this paragraph doesn't look any special. Maybe we should make it more visible. WDYT?

File: docs/admin/cluster-management.md (hunk 1 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: Yes, but it also explains why some nodes that could potentially be removed are not deleted.

Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: Then let's rephrase it to be more vague and say that we will choose "some nodes" that are not needed. I strongly believe that user documentation should include any implementation details, as they will start depending it. It's like with the API.

File: docs/admin/cluster-management.md (hunk 2 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: Done.

File: docs/admin/cluster-management.md (hunk 3 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: Done.

File: docs/admin/cluster-management.md (hunk 4 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: Yikes, still not fixed. Removing.

File: docs/admin/cluster-management.md (hunk 5 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: Done

File: docs/admin/cluster-management.md (hunk 6 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: Done.

File: docs/admin/cluster-management.md (hunk 7 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: Done. 

Roughly = same cpu, mem capacity however the name, ip, possibly zone is different and may run some different pods.

File: docs/admin/cluster-management.md (hunk 1 above)
Comment (mwielgus) [inline docs/admin/cluster-management.md]: I disagree - we can always mark that this strategy may change in the future. 
I wouldn't trust such autoscaler that sometimes may delete some nodes based on some vague conditions.

Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: But then you'd have to describe the whole logic here to be complete. Let's discuss in person/VC tomorrow and ask @piosz and @jszczepkowski for opinion.

File: docs/admin/cluster-management.md (hunk 2 above)
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: s/replciated/replicated

File: docs/admin/cluster-management.md [hunk 8]
```diff
@@ -66,37 +66,56 @@ Instance Group will take care of putting appropriate image on new machines and s
...
 ```shell
-gcloud alpha compute autoscaler --zone $ZONE <command>
+gcloud container clusters update mytestcluster --enable-autoscaling=true --min-nodes=1 --max-nodes=15
 ```
 
-Note that autoscaling will work properly only if node metrics are accessible in Google Cloud Monitoring.
-To make the metrics accessible, you need to create your cluster with `KUBE_ENABLE_CLUSTER_MONITORING`
-equal to `google` or `googleinfluxdb` (`googleinfluxdb` is the default value). Please also make sure
-that you have Google Cloud Monitoring API enabled in Google Developer Console.
+**Cluster autoscaler assumes that all nodes inside a signle instance group or node pool are roughly the same, they don't have any manually added labels, etc. Please create a separate instance group/node pool if you want to add some labels to them.**
```
Comment (fgrzadkowski) [inline docs/admin/cluster-management.md]: s/signle/single

I'd expect that user will the same concerns regarding "roughly the same". If this is only about label we should say this explicitly. Or we can phrase it differently, like "we expect that nodes where not manually modified (e.g. by adding labels via kubectl) as those properties would be propagated to the new nodes within the same instance group".

Comment (mwielgus) [inline docs/admin/cluster-management.md]: Done

**Comment by fgrzadkowski** [comment]
I think that this PR should be against branch `release-1.3`, not master.

**Comment by mwielgus** [comment]
It should also be made in master - the feature will not go away in 1.4. I can create another one for 1.3.

**Comment by fgrzadkowski** [comment]
My understanding is that master is the current release, i.e. 1.2. Later 1.3 will become master, master will become 1.2 and we'll create new branch for 1.4

**Comment by johndmulhausen** [comment]
Yes, this should have been opened against the release-1.3 branch, but you're not alone in the confusion here.

**Comment by fgrzadkowski** [comment]
@johndmulhausen Can we change this PR to be against 1.3 release and merge it now, instead of waiting for the release?

**Comment by erictune** [comment]
Filip: We are asking the PR authors to do the rebase.

To rebase your work on 1.3 do something like this starting in the branch that you used for this PR:
```
COMMITS=$(git log master.. | grep commit | cut -f 2 -d " " | tac)
BR=$(git rev-parse --abbrev-ref HEAD)
git fetch upstream
git checkout upstream/release-1.3
git checkout -b $BR-1.3
for c in $COMMITS
do
  git cherry-pick $c; done
done
git push
```

**Comment by johndmulhausen** [comment]
@fgrzadkowski Yes, if you rebase this against 1.3 we can merge it now! Thanks to @erictune for providing instructions

**Comment by mwielgus** [comment]
Unfortunately it seems that we cannot do this under this PR (or maybe can someone change the target branch of this PR). Please reopen #684 which is for the correct branch.

**Comment by johndmulhausen** [comment]
@mwielgus Reopened PR. Merging that one. Closing this one. :)
